modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/PaginationContextProvider.java

```diff
@@ -22,21 +28,69 @@
 @Provider
 public class PaginationContextProvider implements ContextProvider<Pagination> {
 
+	public PaginationContextProvider(
+		ConfigurationProvider configurationProvider, Portal portal) {
+
+		_configurationProvider = configurationProvider;
+		_portal = portal;
+	}
+
 	@Override
 	public Pagination createContext(Message message) {
 		HttpServletRequest httpServletRequest =
 			ContextProviderUtil.getHttpServletRequest(message);
 
 		String page = httpServletRequest.getParameter("page");
+
+		if ((page != null) && (GetterUtil.getLong(page) <= 0)) {
+			throw new InvalidPaginationException(
+				String.format(
+					"The Page parameter introduced [%s] is not valid. Only " +
+						"numbers higher or equal to 1 are accepted.",
+					page));
+		}
+
 		String pageSize = httpServletRequest.getParameter("pageSize");
+		long pageSizeValueIfNotSpecified = 20L;
+
+		try {
+			HeadlessAPICompanyConfiguration headlessAPICompanyConfiguration =
+				_configurationProvider.getCompanyConfiguration(
+					HeadlessAPICompanyConfiguration.class,
+					_portal.getCompanyId(httpServletRequest));
 
-		if (StringUtil.equals(page, "0") || StringUtil.equals(pageSize, "0")) {
-			return null;
+			Long maxPageSize =
+				headlessAPICompanyConfiguration.getPaginationSizeLimit();
+
+			if ((maxPageSize < 20) && (maxPageSize > 0)) {
+				pageSizeValueIfNotSpecified = maxPageSize;
+			}
+
+			if ((pageSize != null) && (maxPageSize > 0)) {
+				long pageSizeLong = GetterUtil.getLong(pageSize);
+
+				if ((pageSizeLong > maxPageSize) || (pageSizeLong <= 0)) {
+					throw new InvalidPaginationException(
+						String.format(
+							"The PageSize introduced [%s] exceeds the " +
+								"maximum PageSize configured [%s].",
+							pageSize, maxPageSize));
+				}
+			}
+		}
+		catch (ConfigurationException configurationException) {
+			_log.error(configurationException);
 		}
 
 		return Pagination.of(
 			GetterUtil.getInteger(page, 1),
-			GetterUtil.getInteger(pageSize, 20));
+			GetterUtil.getInteger(pageSize, (int)pageSizeValueIfNotSpecified));
```

Comment:
we are passing integers, however, we are dealing with longs in the Configuration, is that correct?

Reply:
Yes, I mainly did it because in an early conversation with you you told me that long values are very used in Liferay, but if you want I can change it to integers

Reply:
yes, but if we are already using integers, it makes no sense to have configuration with long data types

Reply:
Ok, will change it!